### ISSUE TITLE
How do we approach the TypeScript conversion?

### ISSUE THREAD

**Issue by sdepold** [sequelize/meetings#6]
I have just merged #13569

**Comment by AllAwesome497** [comment]
I think we should start with core files then move outwards, otherwise when we convert a core file to TS we might have to change the files which depend on it.

**Comment by sdepold** [comment]
I'm happy with any decision you guys make. Will be happy to merge things since it shouldn't really affect anything (other than making things better)

**Comment by ephys** [comment]
After https://github.com/sequelize/sequelize/pull/13805, @WikiRik and I are starting to think we should postpone the TypeScript migration to v7. We've received a couple of issues regarding unexpected breaking changes in v6.

So far all of them are caused by users accessing the internal APIs (`/lib`). Some TypeScript breakage can also happen if we remove files from `/types` due to IDEs accidentally importing types from the wrong file. I have that issue myself:

<img width="968" alt="Screenshot 2021-12-21 at 18 52 32" src="https://user-images.githubusercontent.com/1280915/146976074-0a95d413-bb23-4e88-9f7d-3d2e5244a5c1.png">

So our options are:

- *option 1*: we consider uses of `/lib` & `/types` to be invalid because they're internal, and breakage in those modules is not considered breaking changes.
- *option 2*: we postpone the migration to v7, where we'll also block importing `/lib` to ensure this does not happen.
- *option 3*: we continue migration in v6, but we try to keep `/lib` & `/types` as they were before the migration (requires hacks https://github.com/sequelize/sequelize/pull/13805/files#diff-38fd6f14c7859313faf73b53c474d7566aca342d74e5338679b752246eff4655L524).

I would opt for *option 2* because *option 3* is a lot of work and *option 1* is already causing users to open issues:

- https://github.com/sequelize/sequelize/issues/13791
- https://github.com/sequelize/sequelize/issues/13803
- https://github.com/sequelize/sequelize/issues/13788

**Comment by sdepold** [comment]
TBH I have no problem at all having the ts changes in v7 and let the people decide if they want to migrate.

**Comment by WikiRik** [comment]
> TBH I have no problem at all having the ts changes in v7 and let the people decide if they want to migrate.

Follow-up (for during the meeting) would be if we want to release v7 as soon as TypeScript migration is done and do other breaking changes in v8 or do we want to still accumulate breaking changes before we release v7?

**Comment by ephys** [comment]
I hope it's ok to put this here before the meeting but: I think it's better to accumulate breaking changes that bring new features or improve existing ones, to give an incentive to upgradde

**Comment by sdepold** [comment]
I guess I would accumulate all the interesting changes and continuously release alpha builds for v7.

**Comment by ephys** [comment]
New issue regarding the migration https://github.com/sequelize/sequelize/issues/13811 :(

**Comment by sdepold** [comment]
Proposal:

0. We focus on v7 now. Aka v6 is completed and people should eventually migrate to v7.
1. Treat `main` as prime development branch for v7 alpha
2. We merge `main` to `v7` every once in a while to release alpha version automatically
3. We semi-manually port important PRs to v6 by either cherry-picking or copying the changes over to the v6 branch (could also be done through PRs targeting v6)
4. Once the TS migration is done, we could release a first beta version of v7 to npm.
5. We introduce `@sequelize/core` afterwards and start splitting the dialects from the core codebase and introduce them as `@sequelize/postgres` (name needs discussion) etc.

What do you think. Happy to have the discussion asynchronously here.

However, I think we can all agree on the fact that we are stepping on everybody's toes when we continue the TS migration on v6. 

Opinion?

**Comment by ephys** [comment]
I vote to do exactly that.

**Comment by WikiRik** [comment]
Agree with the proposal.

Two questions though; 
- Can bigger breaking changes be done in the v7 alphas (so during the TS migration) or do we try to keep those for until after we migrated?
- Splitting modules will be done in the betas of v7 then? Or do we move it to v8? We could use that as promotion since v7 will then really be a big rewrite

**Comment by sdepold** [comment]
Splitting would be part of v7 and we will only turn into 7.0.0 final when every breaking change is in.

**Comment by sdepold** [comment]
or in other words: yes we keep breaking the alpha until we are happy with it :)

**Comment by sdepold** [comment]
FYI: Independently of the outcome of this conversation, I'll prep an alpha pipeline for v7. For that I force replaced the existing v7 branch and stored the former changes in v7-bak (which we can then copy over). With a bit of luck we should have a v7 alpha in a moment: https://github.com/sequelize/sequelize/commit/98ad525ce008b7db3d68985818976fda08e70b67

**Comment by sdepold** [comment]
Status of data types conversion

**Comment by sdepold** [comment]
Sascha: Started working on utils: https://github.com/sequelize/sequelize/pull/13782/files
Zoe: Works on https://github.com/sequelize/sequelize/pull/13922
Wade: https://github.com/sequelize/sequelize/pull/13799

**Comment by ephys** [comment]
Blocking decision: Do we drop support for TypeScript < 4.1?

TypeScript 4.1 added [Key remapping](https://www.typescriptlang.org/docs/handbook/2/mapped-types.html#key-remapping-via-as) in mapped types, and [Template Literal Types](https://www.typescriptlang.org/docs/handbook/2/template-literal-types.html). Both incredibly useful features to improve our typing.

4.1 is also more than a year old now. I think one year of support is enough and dropping < 4.1 would allow us to [greatly reduce the boilerplate](https://github.com/sequelize/sequelize/pull/13909) our typescript users have to go add.

**Comment by WikiRik** [comment]
Feel free to make a PR to drop testing on < 4.1 and add to the docs that we support 4.1 and above and then we'll see if people disagree. But I think a support window for TS versions for at least one year is fine and I would suggest doing a minor version bump whenever we lose support for older versions of TS

**Comment by ephys** [comment]
I agree with the minor bumb. TypeScript itself doesn't follow semver due to the idea that any TypeScript change is a breaking change. I think we should do the same.

**Comment by sdepold** [comment]
As in, any TS version change could contain breaking changes?

**Comment by sdepold** [comment]
FWIW: I'm fine with any solution.

**Comment by ephys** [comment]
Any version of TSC could include a breaking change because as they say it very well themselves, if they fix a bug, something that was not reported as an error in your codebase is now getting reported (https://github.com/microsoft/TypeScript/issues/14116#issuecomment-280592571)

Similarly for us, if we fix our typings / make them stricter (which we've been doing a lot recently), errors that previously went unreported are now getting reported.
Not exactly the same as dropping TS < 4.1 support but I still think we should not include TypeScript in our semver major if the change is reasonable.

**Comment by sdepold** [comment]
Regarding DataType migration: 
- 4 outstanding migrations (which should target the primary data type PR https://github.com/sequelize/sequelize/pull/13799) for MSSQL, SQLITE, MariaDB, DB2 (later DB2 for i)
-

**Comment by sdepold** [comment]
What would be the very next steps afterwards?
- Get utils finished
- Eventually model

New tests should be written in TypeScript (.ts)

**Comment by sdepold** [comment]
Easier associations in TS --> https://github.com/sequelize/meetings/issues/14

**Comment by ephys** [comment]
I've started migrating `AbstractQueryGenerator` to TypeScript, I need to migrate the different query generators to be confident https://github.com/sequelize/sequelize/pull/14020 doesn't break anything.

**Comment by ephys** [comment]
I've also started migrating the associations folder, as it's very often used by `AbstractQueryGenerator`. I may be going down a rabbit hole.

**Comment by sdepold** [comment]
What's still missing right now?

- sequelize.js 
- model.js 
- data types in progress by @AllAwesome497 
- query-generator and association are in progress right now (by @ephys)

**Comment by sdepold** [comment]
data-type PR review was done yesterday.

**Comment by WikiRik** [comment]
Copying @ephys message from Slack here;

To give an idea of where we are at: We have a PR open for the migration of DataTypes, which is an absolutely massive PR (and not close to being finished)
In order of easiest to migrate to hardest, I would say:
- -- easy difficulty --
   - /sql-string.js
   - /model-manager.js
   - /utils/validator-extras.js
   - /instance-validator.js
   - in the dialect folders: all index.js files
      - Depends on almost all other files in the dialect folders, which are complex to type but which can be temporarily typed using .d.ts files
    - in the dialect folders: all query-interface files
       - Main issue is their dependence on query-generator, but we're slowly populating query-generator.d.ts  to fill the gaps
    - most test files
       - It's lower priority but migrating our tests let us find many issues with our current typings
- -- medium difficulty --
   - in the dialect folders: all connection-manager files
   - /hooks.js
   - /sequelize.js
- -- super hard difficulty --
   - data types (in progress). Depends on a lot. Already 161 impacted files.
   - /model.js
     - Huge file
     - This one is going to be insanely hard. Most internal methods should be moved to model-internals.ts and typed first, then model.js  can be migrated
   - in the dialect folders: all query.js files
      - Hard because depends on libraries that are not necessarily implemented. May require splitting the project into a mono-repo first to separate the dialects from core.
   - in the dialect folders: all query-generator.js files
      - Incredibly complex code
      - In the meantime, we are slowly populating query-generator.d.ts files so other files can use it with typings enabled

**Comment by WikiRik** [comment]
Update now the DataTypes PR has been merged;

> In order of easiest to migrate to hardest, I would say:
> - -- easy difficulty --
>    - /model-manager.js
>    - /utils/validator-extras.js
>    - /instance-validator.js
>    - /index.js and /index.mjs
>     - in the dialect folders: all query-interface files
>        - Main issue is their dependence on query-generator, but we're slowly populating query-generator.d.ts  to fill the gaps
>     - most test files
>        - It's lower priority but migrating our tests let us find many issues with our current typings
> - -- medium difficulty --
>    - /sequelize.js
> - -- super hard difficulty --
>    - /model.js
>      - Huge file
>      - This one is going to be insanely hard. Most internal methods should be moved to model-internals.ts and typed first, then model.js  can be migrated
>    - in the dialect folders: all query.js files
>       - Hard because depends on libraries that are not necessarily implemented. May require splitting the project into a mono-repo first to separate the dialects from core.
>    - in the dialect folders: all query-generator.js files
>       - Incredibly complex code
>       - In the meantime, we are slowly populating query-generator.d.ts files so other files can use it with typings enabled

We have also created parent classes for `sequelize.js` and `model.js` so we can migrate them slowly. This approach can also be used for other files

**Comment by WikiRik** [comment]
Update on the TypeScript migration;

- core
  - src/index.js
  - src/index.mjs
  - src/instance-validator.js
  - src/model-manager.js (PR is raised)
  - src/model.js (src/model-typescript.ts exists for slow migration)
  - src/sequelize.js (src/sequelize-typescript.ts exists for slow migration)
  - src/decorators/legacy/index.mjs
  - src/dialects/abstract/query-generator/transaction.js
  - per dialect; src/dialects/*/query.js
  - per dialect; src/dialects/*/query-interface.js (abstract has query-interface-typescript.ts for slow migration)
  - per dialect; src/dialects/*/query-generator.js (each dialect has query-generator-typescript.ts for slow migration)
  - src/utils/validator-extras.js
- validator-js
  - src/index.mjs
- Additional files
  - build-packages.mjs
  - typedoc.js
  - dev/update-authors.js
  - test/register-esbuild.js

**Comment by WikiRik** [comment]
Update on the TypeScript migration;

- core
  - src/index.js
  - src/index.mjs
  - src/instance-validator.js
  - src/model.js (src/model-typescript.ts exists for slow migration)
  - src/sequelize.js (src/sequelize-typescript.ts exists for slow migration)
  - src/decorators/legacy/index.mjs
  - per dialect; src/dialects/**/query.js
  - per dialect except sqlite; src/dialects/**/query-interface.js (abstract, db2 and postgres have query-interface-typescript.ts for slow migration)
  - per dialect; src/dialects/**/query-generator.js (each dialect has query-generator-typescript.ts for slow migration)
  - src/utils/validator-extras.js
  - test/integration/**/*.test.js
  - test/smoke/smoketests.test.js
  - test/unit/**/*.test.js
- validator-js
  - src/index.mjs
- Additional files
  - .eslintrc.js
  - build-packages.mjs
  - typedoc.js
  - dev/update-authors.js
  - test/register-esbuild.js
  - test/esm-named-exports.test.js
